### DISCUSSION TITLE
When customizing a control, how to expose all the properties of an internal control directly to the outside world?

### DISCUSSION THREAD

**Discussion by tsctsc6** [AvaloniaUI/Avalonia#16766] [Q&A] [answered]
For example, I want to create a CustomControls base on DataGrid. By convention, I can create a UserControl or TemplatedControl, and combine some controls. But I don't want to lose the features of the DataGrid, like AutoGenerateColumns, IsReadOnly, DataGrid.Columns, etc. Of course, I could redefine and map the corresponding properties to the internal DataGrid one by one, but it's a bit of a pain. Is there a more convenient way?

**Comment by stevemonaco** [comment, marked as answer]
If you are deriving from `UserControl` or `TemplatedControl`, you need to manually forward each property. Otherwise, what if your `UserControl` contained two `TextBox`? There would be duplicate properties.

IMO, the more controls you try to componentize together, the less reusable your control actually is. There may be some things the team can do for `UserControl`s, but AFAIK, it would need to be designed from scratch as other XAML frameworks don't have a convenient forwarding mechanism to look at as prior art.

You probably should instead consider deriving directly from `DataGrid` and create a new `ControlTheme` based upon your theme's. ie. [Fluent DataGrid](https://github.com/AvaloniaUI/Avalonia/blob/2d516027602698b1aa817f46edc62fd9d337c66a/src/Avalonia.Controls.DataGrid/Themes/Fluent.xaml#L500-L578). Since you own the `ControlTemplate`, you can easily change the XAML structure to include other controls (properties that you'll need to manually forward, if actually necessary).

**Comment by tsctsc6** [comment]
You've got a point there. And I try to deriving directly from `TextBox` , has a weird error "Required template part with x:Name 'PART_TextPresenter' must be defined on 'TemplatedControl2' ControlTemplate.".

Maybe I can use Source Generators, which auto generate properties by some naming convention, but it's a little complicated to write. So I'll just do it manually.

**Reply by stevemonaco** [reply]
ControlThemes typically have very loose requirements, but controls can require that their `ControlTemplate` includes controls vital for its operation. So the custom `TextBox` you've derived must also include the `PART_TextPresenter` in its `ControlTemplate`.

Aside from that, maybe you didn't define a `ControlTemplate` at all and were expecting Avalonia to automatically use the default for `TextBox`. This is not the case. [See the docs](https://docs.avaloniaui.net/docs/basics/user-interface/styling/styles#style-key) for how to use `StyleKeyOverride`.

**Reply by tsctsc6** [reply]
So, in this way, I still have to define some property or something else which exist in base class. That's not what I was going for. I didn't want to define them one by one, right?

**Reply by stevemonaco** [reply]
This is separate from your want to forward properties in a class. If a `TemplatedControl` you derive from requires something, you must put it into the `ControlTemplate` (defined in XAML, not a class). Which is something you're going to have to create anyways if you're trying to mitigate the number of properties you need to forward by deriving.

**Reply by tsctsc6** [reply]
I see. Now I can deriving directly from `TextBox` and show it on window. But this way doesn't seem to be possible to add other controls, such as a button.

**Reply by workgroupengineering** [reply]
TextBox and some other controls have 
InnerLeftContent and InnerRightContent. You can use like this

```xml
<TextBox>
  <TextBox.InnerRightContent>
    <Button Content="..."/>
   </TextBox.InnerRightContent>
</TextBox>
```

**Comment by tsctsc6** [comment]
However, there another question here. Look the code below:
```
<MyControl ItemsSource="{Binding People}">
  <MyControl.DataGridColumns>
     <DataGridTextColumn Header="First Name"  Binding="{Binding FirstName}"/>
     <DataGridTextColumn Header="Last Name" Binding="{Binding LastName}" />
  </MyControl.DataGridColumns>
</MyControl>
```
Suppose there is a `DataGrid` in `MyControl` . How to define `MyControl.DataGridColumns` ? How to sent `DataGridTextColumn` to the inner `DataGrid`?

**Reply by stevemonaco** [reply]
Probably the same way `DataGridColumns` is defined in `DataGrid`. You can read through the Avalonia themes to see examples of forwarding in most controls as they forward `Foreground`, `Background`, or border properties into an internal `<Border>` or `<ContentPresenter>`. Though I'm not 100% sure you can forward collections as I avoid componentization unless there's a good justification as I mentioned earlier.

Then you use a `<DataGrid DataGridColumns="{TemplateBinding DataGridColumns}">` assuming you've created a `ControlTemplate`. Otherwise, use an indirect binding such as `ElementName` or `RelativeSource`.

**Reply by tsctsc6** [reply]
I see. But I'm not understand the source code.

**Reply by tsctsc6** [reply]
I can't find keyword "Columns" in source code

**Comment by tsctsc6** [comment]
For the properties of a custom control, I see three types in the [example](https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/CustomControls/RatingControlSample#properties). But there is no introduction of properties like `DataGrid.Columns` .

**Reply by stevemonaco** [reply]
After checking the source for `DataGrid.Columns`, it is more complex because it's actually not a dependency property. In the case where you're manually defining columns, this is just adding items via fetching the collection. When it's autogenerated, it's populated by code. There's no binding involved for this particular property.

https://github.com/AvaloniaUI/Avalonia/blob/5581516e6b091906831c0398bcca8897f651d084/src/Avalonia.Controls.DataGrid/DataGrid.cs#L1370-L1378

So you may need to manually synchronize here or maybe create a required template part with `DataContext`, expose a get-only `Columns` property like `DataGrid` does, fetch the template part, and return `dataGrid.Columns`.

This is why I say to derive from `DataGrid` if most of the functionality needs forwarded anyways.

**Reply by tsctsc6** [reply]
Thanks for helping me find the definition of Columns. As you said before, "the more controls you try to componentize together, the less reusable your control actually is.". So I think it's fine to just implement the necessary features. In other words, it's fine to just manual forward the necessary properties.

**Reply by stevemonaco** [reply]
The problem is identifying the "necessary" properties. To expand onto the advice: the fewer properties you forward (as an overall percentage), the less reusable it actually is. Once you start adding many child controls, it's also hard to decide which should have properties like `HorizontalAlignment` or `Background` applied. If you're creating a one-time use control, you might as well hard-code/style the properties instead of forwarding.

**Reply by tsctsc6** [reply]
I see, thank you.

**Comment by tsctsc6** [comment]
Let me summarize.

First, the way of combining controls. For me, it is easier to write, but I need to manual forward properties.

Second, the way of inherite control. Although it doesn't require me to manually forward the properties, but it creates more problems, and complexity in writing code. For example, I can't combining controls in this way(some controls don't have
InnerLeftContent and InnerRightContent). And it's not as flexible as combining controls.

Thanks for your help.
